example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+
+                std::cout << "* SOCKS remote host: " << domain << ":" << port << std::endl;
+            }
+
+            if (atyp == detail::SOCKS5_ATYP_IPV4)
+            {
+                net::ip::tcp::endpoint remote_endp(
+                    net::ip::address_v4(read<uint32_t>(resp)),
+                    read<uint16_t>(resp));
+
+                std::cout << "* SOCKS remote host: " << remote_endp.address().to_string()
+                    << ":" << remote_endp.port() << std::endl;
+            }
+
+            if (atyp == detail::SOCKS5_ATYP_IPV6)
+            {
+                net::ip::address_v6::bytes_type bytes;
+                for (auto i = 0; i < 16; i++)
+                    bytes[i] = read<uint8_t>(resp);
+
+                net::ip::tcp::endpoint remote_endp(
+                    net::ip::address_v6(bytes),
+                    read<uint16_t>(resp));
+
+                std::cout << "* SOCKS remote host: " << remote_endp.address().to_string()
+                    << ":" << remote_endp.port() << std::endl;
+            }
+
+            // fail.
+            if (rep != 0)
+            {
+                switch (rep)
+                {
+                case detail::SOCKS5_GENERAL_SOCKS_SERVER_FAILURE:
+                    ec = error::socks_general_failure; break;
+                case detail::SOCKS5_CONNECTION_NOT_ALLOWED_BY_RULESET:
+                    ec = error::socks_connection_not_allowed_by_ruleset; break;
+                case detail::SOCKS5_NETWORK_UNREACHABLE:
+                    ec = error::socks_network_unreachable; break;
+                case detail::SOCKS5_CONNECTION_REFUSED:
+                    ec = error::socks_connection_refused; break;
+                case detail::SOCKS5_TTL_EXPIRED:
+                    ec = error::socks_ttl_expired; break;
+                case detail::SOCKS5_COMMAND_NOT_SUPPORTED:
+                    ec = error::socks_command_not_supported; break;
+                case detail::SOCKS5_ADDRESS_TYPE_NOT_SUPPORTED:
+                    ec = error::socks_address_type_not_supported; break;
+                default:
+                    ec = error::socks_unassigned; break;
+                }
+
+                break;
+            }
+            BOOST_FALLTHROUGH;
+        }
+        case 10:
+            break;
+        }
+
+        this->complete_now(ec);
+    }
+
+private:
+    Stream& stream_;
+
+    using BufferPtr = std::unique_ptr<Buffer>;
+    BufferPtr request_ { new Buffer() };
+    BufferPtr response_ { new Buffer() };
+
+    std::string hostname_;
+    unsigned short port_;
+    std::string username_;
+    std::string password_;
+    bool use_hostname_;
+    int step_ = 0;
+};
+
+/** Perform the SOCKS v4 handshake in the client role.
+*/
+template<
+    typename AsyncStream,
+    typename Handler
+>
+BOOST_BEAST_ASYNC_RESULT1(Handler)
+async_handshake_v4(
+    AsyncStream& stream,
+    const std::string& hostname,
+    unsigned short port,
+    std::string const& username,
```

Comment:
Should use `string_view` instead of `std::string`, and throw if the protocol length limits are exceeded (e.g. `hostname.size() > 255`)